protocols/relay/src/protocol/outbound_stop.rs

```diff
@@ -161,3 +73,152 @@ pub enum FatalUpgradeError {
     #[error("Unexpected message status '{0:?}'")]
     UnexpectedStatus(proto::Status),
 }
+
+pub(crate) async fn handle_stop_message_response(
+    io: Stream,
+    stop_command: StopCommand,
+    tx: oneshot::Sender<()>,
+) -> handler::RelayConnectionHandlerEvent {
+    let msg = proto::StopMessage {
+        type_pb: proto::StopMessageType::CONNECT,
+        peer: Some(proto::Peer {
+            id: stop_command.src_peer_id.to_bytes(),
+            addrs: vec![],
+        }),
+        limit: Some(proto::Limit {
+            duration: Some(
+                stop_command
+                    .max_circuit_duration
+                    .as_secs()
+                    .try_into()
+                    .expect("`max_circuit_duration` not to exceed `u32::MAX`."),
+            ),
+            data: Some(stop_command.max_circuit_bytes),
+        }),
+        status: None,
+    };
+
+    let mut substream = Framed::new(io, quick_protobuf_codec::Codec::new(MAX_MESSAGE_SIZE));
+
+    let send_res = substream.send(msg).await;
+    if send_res.is_err() {
+        return ConnectionHandlerEvent::Close(StreamUpgradeError::Apply(Either::Right(
+            FatalUpgradeError::StreamClosed,
+        )));
+    }
+
+    let res = substream.next().await;
+
+    if let None | Some(Err(_)) = res {
+        return ConnectionHandlerEvent::Close(StreamUpgradeError::Apply(Either::Right(
+            FatalUpgradeError::StreamClosed,
+        )));
+    }
+
+    let proto::StopMessage {
+        type_pb,
+        peer: _,
+        limit: _,
+        status,
+    } = res.unwrap().expect("should be ok");
+
+    match type_pb {
+        proto::StopMessageType::CONNECT => {
+            return ConnectionHandlerEvent::Close(StreamUpgradeError::Apply(Either::Right(
+                FatalUpgradeError::UnexpectedTypeConnect,
+            )))
+        }
+        proto::StopMessageType::STATUS => {}
+    }
+
+    match status {
+        Some(proto_status) => match proto_status {
+            proto::Status::OK => {}
```

Comment:
These `match` can be nested which will reduce the level of indentation here which makes this easier to read.

```suggestion
    match status {
        Some(proto::Status::OK) => {}
```